numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -259,6 +254,10 @@ PyArray_AdaptFlexibleDType(PyObject *data_obj, PyArray_Descr *data_dtype,
                              * GetArrayParamsFromObject won't iterate over
                              * array.
                              */
+                            PyArray_Descr *dtype = NULL;
+                            PyArrayObject *arr = NULL;
+                            int result, ndim = 0;
+                            npy_intp dims[NPY_MAXDIMS];
                             list = PyArray_ToList((PyArrayObject *)data_obj);
                             result = PyArray_GetArrayParamsFromObject(
```

Comment:
Seems there's another bug here - this doesn't propagate exceptiosn correctly when `result < 0`

Reply:
In fact, `PyArray_AdaptFlexibleDType` really ought to have an `int` return value

Reply:
Out of scope for this PR though

Reply:
see issue #11969